aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+        ObjectId { inner: object_id }
+    }
+
+    /// Derives an object id from source material: sha3_256([creator address | seed | 0xFE]).
+    /// The ObjectId needs to be distinct from create_resource_address
+    public fun derive_object_id(source: &address, seed: vector<u8>): ObjectId {
+        let bytes = bcs::to_bytes(source);
+        vector::append(&mut bytes, seed);
+        vector::push_back(&mut bytes, OBJECT_ADDRESS_SCHEME);
+        ObjectId { inner: from_bcs::to_address(hash::sha3_256(bytes)) }
+    }
+
+
+    /// Returns the address of an ObjectId
+    public fun object_id_address(object_id: &ObjectId): address {
+        object_id.inner
+    }
+
+    /// Create a new named object and return the CreatorRef. Named objects can be queried globally
+    /// by knowing the user generated seed used to create them. Named objects cannot be deleted.
+    public fun create_named_object(creator: &signer, seed: vector<u8>): CreatorRef {
+        let creator_address = signer::address_of(creator);
+        let id = derive_object_id(&creator_address, seed);
+        create_object_internal(creator_address, id, false)
+    }
+
+    public fun create_object_from_account(creator: &signer): CreatorRef {
+        let guid = account::create_guid(creator);
+        create_object_from_guid(signer::address_of(creator), guid)
+    }
+
+    public fun create_object_from_object(creator: &signer): CreatorRef acquires Object {
+        let guid = create_guid(creator);
+        create_object_from_guid(signer::address_of(creator), guid)
+    }
+
+    fun create_object_from_guid(creator_address: address, guid: guid::GUID): CreatorRef {
+        let bytes = bcs::to_bytes(&guid);
+        vector::push_back(&mut bytes, OBJECT_ADDRESS_SCHEME);
+        let object_id = ObjectId { inner: from_bcs::to_address(hash::sha3_256(bytes)) };
+        create_object_internal(creator_address, object_id, true)
+    }
+
+    fun create_object_internal(
+        creator_address: address,
+        id: ObjectId,
+        can_delete: bool,
+    ): CreatorRef {
+        assert!(!exists<Object>(id.inner), error::already_exists(EOBJECT_EXISTS));
+
+        let object_signer = create_signer(id.inner);
+        let guid_creation_num = 0;
+        let transfer_events_guid = guid::create(id.inner, &mut guid_creation_num);
+
+        move_to(
+            &object_signer,
+            Object {
+                guid_creation_num,
+                owner: creator_address,
+                allow_ungated_transfer: true,
+                transfer_events: event::new_event_handle(transfer_events_guid),
+            },
+        );
+        CreatorRef { self: id, can_delete }
+    }
+
+    // Creation helpers
+
+    public fun disallow_ungated_transfer(ref: &CreatorRef) acquires Object {
+        let object = borrow_global_mut<Object>(ref.self.inner);
+        object.allow_ungated_transfer = false;
+    }
+
+    public fun generate_delete_ref(ref: &CreatorRef): DeleteRef {
```

Comment:
could people want to delete after not processing creatorRef anymore?

Reply:
Someone might always want more than what they asked for initially, I think we should be opinionated that some decisions are only available once... If they want to make a framework that exposes DeleteRef for later, more power to them.

Reply:
Not sure if I miss read the code. But it seems the owner of the object may not possess the creatorRef and won't have a way to delete.

Reply:
I specified in the AIP that decisions on the evolution of the object must be defined at creation... you have unlimited flexibility to state you want unlimited flexibility -- at creation time.